research/picai/preprocessing.py

```diff
@@ -0,0 +1,381 @@
+from __future__ import annotations
+
+import os
+from abc import ABC, abstractmethod
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from functools import partial, reduce
+from pathlib import Path
+from typing import List, Optional, Sequence, Tuple
+
+import numpy as np
+import SimpleITK as sitk
+from preprocessing_transforms import crop_or_pad, resample_img
+
+
+@dataclass
+class PreprocessingSettings:
+    def __init__(
+        self,
+        scans_write_dir: Path,
+        annotation_write_dir: Path,
+        size: Optional[Tuple[int, int, int]],
+        physical_size: Optional[Tuple[float, float, float]],
+        spacing: Optional[Tuple[float, float, float]],
+    ) -> None:
+        """
+        Dataclass encapsulating parameters of preprocessing.
+
+        Args:
+            scans_write_dir (Path): The directory to write the preprocessed scans.
+            annotation_write_dir (Path): The directory to write the preprocessed annotation.
+            size (Optional[Tuple[int, int, int]]): Tuple of 3 int representing size of scan in voxels.
+                In the format of Depth x Height x Width. If None, preprocessed scans and annotations retain
+                their original size.
+            physical_size (Optional[Iterable[float]]): Tuple of 3 float representing actual size of scan in mm.
+                In the format of Depth x Height x Width. If None and size and spacing are not None,
+                physical_size will be inferred.
+            spacing (Optional[Sequence[float]]): Tuple of 3 float representing spacing between voxels
+                of scan in mm/voxel. In the format of Depth x Height x Width. If None,
+                preprocessed scans and annotations retain their original spacing.
+        """
+        self.scans_write_dir = scans_write_dir
+        self.annotation_write_dir = annotation_write_dir
+        self.size = size
+        self.spacing = spacing
+
+        if physical_size is None and self.spacing is not None and self.size is not None:
+            # calculate physical size
+            self.physical_size = (
```

Comment:
Since these are tuple[int, int, int] now, I had to unpack to ensure the typing works